jupyterlab/coreconfig.py

```diff
@@ -0,0 +1,168 @@
+            data['resolutions'] = {}
+            data['jupyterlab']['extensions'] = {}
+            data['jupyterlab']['mimeExtensions'] = {}
+            data['jupyterlab']['singletonPackages'] = []
+
+    def add(self, name, semver, extension=False, mime_extension=False):
+        """Remove an extension/singleton.
+
+        If neither extension or mimeExtension is True (the default)
+        the package is added as a singleton dependency.
+
+        name: string
+            The npm package name
+        semver: string
+            The semver range for the package
+        extension: bool
+            Whether the package is an extension
+        mime_extension: bool
+            Whether the package is a MIME extension
+        """
+        data = self._data
+        if not name:
+            raise ValueError('Missing package name')
+        if not semver:
+            raise ValueError('Missing package semver')
+        if name in self._data['resolutions']:
+            raise ValueError('Package already present: %r' % (name,))
+        data['resolutions'][name] = semver
+
+        # If both mimeExtension and extensions are True, treat
+        # as mime extension
+        if mime_extension:
+            data['jupyterlab']['mimeExtensions'][name] = ""
+            data['dependencies'][name] = semver
+        elif extension:
+            data['jupyterlab']['extensions'][name] = ""
+            data['dependencies'][name] = semver
+        else:
+            data['jupyterlab']['singletonPackages'].append(name)
+
+    @property
+    def extensions(self):
+        """A dict mapping all extension names to their semver"""
+        return dict(
+            (k, self._data['resolutions'][k])
+            for k in self._data['jupyterlab']['extensions'].keys())
+
+    @property
+    def mime_extensions(self):
+        """A dict mapping all MIME extension names to their semver"""
+        return dict(
+            (k, self._data['resolutions'][k])
+            for k in self._data['jupyterlab']['mimeExtensions'].keys())
+
+    @property
+    def singletons(self):
+        """A dict mapping all singleton names to their semver"""
+        return dict(
+            (k, self._data['resolutions'][k])
+            for k in self._data['jupyterlab']['singletonPackages']
+        )
+
+    def remove(self, name):
+        """Remove a package/extension.
+
+        name: string
+            The npm package name
+        """
+        data = self._data
+        maps = (
+            data['dependencies'],
+            data['resolutions'],
+            data['jupyterlab']['extensions'],
+            data['jupyterlab']['mimeExtensions'],
+        )
+        for m in maps:
+            try:
+                del m[name]
+            except KeyError:
+                pass
+
+        data['jupyterlab']['singletonPackages'].remove(name)
+
+
+    def set_static_dir(self, static_dir):
+        self._data['jupyterlab']['staticDir'] = static_dir
+
+    @property
```

Comment:
doesn't there need to be standard structure here?

Reply:
Not sure what you mean